pkg/otel/arrow_record/consumer.go

```diff
@@ -94,20 +135,86 @@ type Option func(*Config)
+		memLimit:      defaultMemoryLimit,
+		tracesConfig:  arrow.DefaultConfig(),
+		meterProvider: otel.GetMeterProvider(),
 	}
 	for _, opt := range opts {
 		opt(&cfg)
 	}
+	meter := cfg.meterProvider.Meter("otel-arrow/pkg/otel/arrow_record")
+	allocator := common.NewLimitedAllocator(memory.NewGoAllocator(), cfg.memLimit)
 	return &Consumer{
-		config:          cfg,
-		streamConsumers: make(map[string]*streamConsumer),
+		Config:             cfg,
+		allocator:          allocator,
+		uniqueAttr:         attribute.String("stream_unique", fmt.Sprintf("%08x", rand.Uint32())),
+		streamConsumers:    make(map[string]*streamConsumer),
+		recordsCounter:     mustWarn(meter.Int64Counter("arrow_batch_records")),
+		schemaResetCounter: mustWarn(meter.Int64Counter("arrow_schema_resets")),
+		memoryCounter:      mustWarn(meter.Int64UpDownCounter("arrow_memory_inuse")),
+	}
+}
+
+func releaseRecords(recs []*record_message.RecordMessage) {
+	for _, rec := range recs {
+		rec.Record().Release()
+	}
+}
+
+func mustWarn[T any](t T, err error) T {
+	if err != nil {
+		// as it's an otel error, let someone else handle it
+		otel.Handle(err)
+	}
+	return t
+}
+
+type placeholder struct{}
+
+// inuseChangeByCaller records the change in allocated memory,
+// attributing change to the caller.  The `placeholder` returned
+// is to encourage this idiom at the top of any function that
+// allocates memory in the library.
+//
+//	defer c.inuseChangeObserve(c.inuseChangeByCaller())
+//
+// Note: This is a sanity check. Currently there are no allocator
+// operations in the caller, and the caller is not responsible
+// for releasing memory. The count by where=caller should equal 0.
+func (c *Consumer) inuseChangeByCaller() placeholder {
+	c.inuseChangeObserveWhere("caller")
+	return placeholder{}
+}
+
+// inuseChangeObserve records the change in allocated memory,
+// attributing change to the library.
+//
+// Note if we have a memory accounting leak, we expect it will
+// show up as a count `arrow_memory_inuse{where=library}`.
+func (c *Consumer) inuseChangeObserve(_ placeholder) {
+	c.inuseChangeObserveWhere("library")
+}
+
+// inuseChangeObserveWhere records synchronous UpDownCounter events
+// tracking changes in allocator state.  If OTel had a synchronous
+// cumulative updowncounter option, that would be easier to use.
+func (c *Consumer) inuseChangeObserveWhere(where string) {
+	ctx := context.Background()
+	last := c.lastInuseValue
+	inuse := c.allocator.Inuse()
+	attrs := metric.WithAttributes(c.uniqueAttr, attribute.String("where", where))
+
+	c.memoryCounter.Add(ctx, int64(inuse-last), attrs)
+	c.lastInuseValue = inuse
+	if inuse != last {
+		fmt.Println("change by", where, "=", int64(inuse-last), "; current value", inuse)
 	}
 }
 
 // MetricsFrom produces an array of [pmetric.Metrics] from a BatchArrowRecords message.
 func (c *Consumer) MetricsFrom(bar *colarspb.BatchArrowRecords) ([]pmetric.Metrics, error) {
+	defer c.inuseChangeObserve(c.inuseChangeByCaller())
```

Comment:
basic question: what is the purpose of nesting these calls? i.e. what is the reason for returning `placeholder{}` and what is the difference between this and function signatures that allow sequential defers like
```
defer c.unuseChangeObserve()
defer c.inuseChangeByCaller()
```

Reply:
See https://go.dev/doc/effective_go#defer for some detail.

> The arguments to the deferred function (which include the receiver if the function is a method) are evaluated when the defer executes, not when the call executes.

This idiom allows me to use 1 line. The equivalent would be:

```
c.inuseChangeByCaller()
defer c.inuseChangeObserve()
```

if after learning about defer's evaluation order you think that's more readable, I'll change it!